internal/prune/graphprune.go

```diff
@@ -48,143 +44,55 @@ func Snip(v1 *viper.Viper, mc *minio.Client) error {
+			g, err := graph.MakeURN(oa[x], bucketName)
+			if err != nil {
+				log.Error("gets3Bytes %v\n", err)
+				// should this just return. since on this error things are not good
+			}
+			oag = append(oag, g)
 		}
 
-		//compare lists..   anything IN graph not in objects list should be removed
-		d := difference(ga, oa) // return array of items in ga that are NOT in oa
+		//compare lists, anything IN graph not in objects list should be removed
+		d := difference(ga, oag) // return array of items in ga that are NOT in oa
+		m := findMissingElements(oag, ga)
 
-		fmt.Printf("Graph items: %d  Object items: %d  difference: %d\n", len(ga), len(oa), len(d))
+		fmt.Printf("Graph items: %d  Object items: %d  difference: %d\n", len(ga), len(oag), len(d))
+		fmt.Printf("Missing item count: %d\n", len(m))
 
 		// For each in d will delete that graph
-		bar := progressbar.Default(int64(len(d)))
-		for x := range d {
-			log.Printf("Remove graph: %s\n", d[x])
-			sparqlapi.Drop(v1, d[x])
-			bar.Add(1)
+		if len(d) > 0 {
+			bar := progressbar.Default(int64(len(d)))
+			for x := range d {
+				log.Printf("Remove graph: %s\n", d[x])
+				graph.Drop(v1, d[x])
+				bar.Add(1)
+			}
 		}
-	}
-
-	return nil
-}
-
-func graphList(v1 *viper.Viper, mc *minio.Client, prefix string) ([]string, error) {
-	ga := []string{}
-
-	//spql := v1.GetStringMapString("sparql")
-	//objs := v1.GetStringMapString("objects")
-	spql, _ := config.GetSparqlConfig(v1)
-	//objs,_ := config.GetObjectsConfig(v1)
-	bucketName, _ := config.GetBucketName(v1)
-	//gp := fmt.Sprintf("urn:%s:%s", objs["bucket"], strings.Replace(prefix, "/", ":", -1))
-	gp := fmt.Sprintf("urn:%s:%s", bucketName, strings.Replace(prefix, "/", ":", -1))
-	fmt.Printf("Pattern: %s\n", gp)
-
-	d := fmt.Sprintf("SELECT DISTINCT ?g WHERE {GRAPH ?g {?s ?p ?o} FILTER regex(str(?g), \"^%s\")}", gp)
-
-	//fmt.Println(d)
-
-	pab := []byte("")
-	params := url.Values{}
-	params.Add("query", d)
-	//req, err := http.NewRequest("GET", fmt.Sprintf("%s?%s", spql["endpoint"], params.Encode()), bytes.NewBuffer(pab))
-	req, err := http.NewRequest("GET", fmt.Sprintf("%s?%s", spql.Endpoint, params.Encode()), bytes.NewBuffer(pab))
-	if err != nil {
-		log.Println(err)
-	}
-	req.Header.Set("Accept", "application/sparql-results+json")
-
-	//req.Header.Add("Accept", "application/sparql-update")
-	//req.Header.Add("Accept", "application/n-quads")
 
-	client := &http.Client{}
-	resp, err := client.Do(req)
-	if err != nil {
-		log.Println(err)
-	}
-
-	defer resp.Body.Close()
+		// load new ones..
+		spql, err := config.GetSparqlConfig(v1)
+		if err != nil {
+			log.Error("prune -> config.GetSparqlConfig %v\n", err)
+		}
 
-	body, err := ioutil.ReadAll(resp.Body)
-	if err != nil {
-		log.Println(strings.Repeat("ERROR", 5))
-		log.Println("response Status:", resp.Status)
-		log.Println("response Headers:", resp.Header)
-		log.Println("response Body:", string(body))
+		if len(m) > 0 {
+			bar2 := progressbar.Default(int64(len(d)))
```

Comment:
m?

Reply:
So prune needs to both the difference (d) and missing (m)

d are items in the graph currently that are no longer in the object store (harvest).  These are removed from the graph

m is items not in the graph but in the object store.  So new items in the harvest that have been added.  These are loaded to the graph.  

did I mess up the logic in the code elsewhere?

Reply:
at this point, d is undefined, so progressbar bar2. always shows zero.
Changed the bar2 to m,  bar2 := progressbar.Default(int64(len(m)))
Progress bar increments.

Add log.Trace , and log the 'statistics' using fields, so that if run into something like splunk we can pull them out.